### DISCUSSION TITLE
How to create multiple services using opentelemetry (java) sdk?

### DISCUSSION THREAD

**Discussion by MacNale** [open-telemetry/opentelemetry-java#2987] [Q&A]
I have an application which is made up of multiple services built in proprietary language. I want to collect traces and ingest into Jaeger or APM solution. There is no instrumentation library. However, these services produce traces in a proprietary format. I want to convert these traces to OpenTelemetry traces and ingest into Jaeger or APM solution.

I started using OpenTelemetry-Java SDK.

```
{
OpenTelemetry openTelemetry = ExampleConfiguration.initOpenTelemetry();
Tracer tracer = openTelemetry.getTracer("io.opentelemetry.example");
Span exampleSpan = tracer.spanBuilder("exampleSpan").startSpan();
childOne(exampleSpan, openTelemetry);
...
}

 private static void childOne(Span parentSpan, OpenTelemetry openTelemetry) {
    Tracer tracer = openTelemetry.getTracer("service2");
    Span childSpan = tracer.spanBuilder("child").setParent(Context.current().with(parentSpan)).startSpan();
    try {
      Thread.sleep(4000);
    } catch (InterruptedException e) {
      // TODO Auto-generated catch block
      e.printStackTrace();
    }
    childSpan.end();
  }
```
Configured appropriate exporter. I can see the traces in Jaeger as expected.
![image](https://user-images.githubusercontent.com/14235758/110075343-485b6a80-7d48-11eb-9f4a-6c94fafbc110.png)

BUT it shows only one service.
![image](https://user-images.githubusercontent.com/14235758/110075377-590be080-7d48-11eb-8753-d379a811a692.png)

Whereas I want to depict two different services. Is that possible using OpenTelemetry?

Here is what I want to achieve -
![image](https://user-images.githubusercontent.com/14235758/110131615-f2aab080-7d8f-11eb-81ed-f7ed5bdf4455.png)

**Comment by anuraaga** [comment]
Hi @MacNale - Can you describe your use case more? In your example, both spans are generated in the same application, which is considered a service in the Jaeger sense so having the same service name makes sense. Note that this service is the local service, not remote one. Even if your app talks to many different remote services, the spans in the app would all have the same local service name. Are you trying to achieve something different?

**Reply by MacNale** [reply]
Hi @anuraaga, I updated the question with the conceptual diagram. Is that helpful?

Basically, this is an out of band spans stitched together in a converter service.

**Reply by jkwatson** [reply]
Could you configure two separate SDK instances, each with their own Resource with the appropriate `service.name` and have your "Converter service" choose the appropriate one to use?

**Reply by MacNale** [reply]
@jkwatson 

I created two OpenTelemetry objects and used to depict "ParentService" and "ChildService". 

```
    OpenTelemetry openTelemetryParent = ExampleConfiguration.initOpenTelemetry(jaegerHostName, jaegerPort, "ParentService" );
    Tracer tracerParent = openTelemetryParent.getTracer("FatherServiceAgent");
    
    OpenTelemetry openTelemetryChild = ExampleConfiguration.initOpenTelemetry(jaegerHostName, jaegerPort, "ChildService" );
    Tracer tracerChild = openTelemetryChild.getTracer("ChildServiceAgent");
    
    for (int i = 0; i < 10; i++) {
      Span exampleSpan = tracerParent.spanBuilder("MyService1Span").startSpan();
      exampleSpan.setAttribute(ResourceAttributes.SERVICE_NAME, "MyService1");
      
      try (Scope scope = exampleSpan.makeCurrent()) {
      
        jaegerchildOne(exampleSpan, tracerChild);
        
        Thread.sleep(100);
      } catch (InterruptedException e) {
        e.printStackTrace();
      } finally {
        exampleSpan.end();
      }
    }
  }
```

I see two services in Jaeger -
![image](https://user-images.githubusercontent.com/14235758/110190077-2e249980-7de7-11eb-99bf-5caf64207ce3.png)

1) Is it okay to create two or more OpenTelemetry Objects in a single service as above?
2) JaegerGrpcSpanExporter is used as exporter. Will it open multiple communication channels to the backend?
3) Are there any gotchas, I should watch for?

**Reply by jkwatson** [reply]
1. It is 100% ok to do this. They're your resources, do with them as you will.  :)
2. If you have two instances of the exporter, then yes, you will have multiple connections. Sharing an exporter is possible, but I don't think it's been tested extensively, so be careful with that, if you try it.
3. See my comments on 2) about sharing an exporter.

**Reply by MacNale** [reply]
@jkwatson  Many Thanks!!

Just wondering if there is any other way to solve my problem. I was advised to use "Resource". I wish if I could use Resource as follows - 

```
private static void firstService(Span parentSpan, Tracer tracer) {
   Span exampleSpan = tracer.spanBuilder("MyService1Span").startSpan();
   exampleSpan.setAttribute(ResourceAttributes.SERVICE_NAME, "MyService1");

   secondService(exampleSpan, tracer);
}

private static void secondService(Span parentSpan, Tracer tracer) {
    Span childSpan = tracer.spanBuilder("MyService2Span").setParent(Context.current().with(parentSpan)).startSpan();
    childSpan.setAttribute(ResourceAttributes.SERVICE_NAME, "MyService2");
}
```

**Reply by jkwatson** [reply]
Unfortunately, you're not actually using a `Resource` in that example, but just grabbing a standard attribute for a Resource and putting it onto the Spans. The Resource is associated with the SdkTracerProvider, so all spans that are generated via Tracers from that provider will share a resource.

**Reply by MacNale** [reply]
@jkwatson 

Updating this thread, to help others having the same problem - 

Created two or more providers in the 'convertor service'.  Created only one exporter and reused to create spanprocessor for each provider.

```
OpenTelemetry openTelemetryParent = ExampleConfiguration.initOpenTelemetry(jaegerHostName, jaegerPort,
        "FirstService");
    Tracer tracerParent = openTelemetryParent.getTracer("FirstServiceInstrumentation");

    OpenTelemetry openTelemetryChild = ExampleConfiguration.initOpenTelemetry(jaegerHostName, jaegerPort,
        "SecondService");
    Tracer tracerChild = openTelemetryChild.getTracer("SecondServiceInstrumentation");

    Random randomNum = new Random();

    for (int i = 0; i < 10; i++) {
      Span parentSpan = tracerParent.spanBuilder("MyService1Span").startSpan();
      try (Scope scope = parentSpan.makeCurrent()) {
        Thread.sleep(1 + randomNum.nextInt(100));
        secondService(parentSpan, tracerChild, randomNum.nextInt(300));
        Thread.sleep(1 + randomNum.nextInt(100));
      } catch (InterruptedException e) {
        e.printStackTrace();
      } finally {
        parentSpan.end();
      }
    }
  }

  private static void secondService(Span parentSpan, Tracer tracer, int delay) {
    Span childSpan = tracer.spanBuilder("MyService2Span").setParent(Context.current().with(parentSpan)).startSpan();
    try {
      Thread.sleep(delay);
    } catch (InterruptedException e) {
      e.printStackTrace();
    }
    childSpan.end();
  }

```

I see two services in Jaeger as expected.
